gdal/alg/marching_squares/polygon_ring_appender.h

```diff
@@ -51,59 +52,46 @@ class PolygonRingAppender
         Ring& operator=( const Ring& other ) = default;
 
         LineString points;
-        
-        mutable std::list<const Ring*> interiorRings;
+
+        mutable std::vector<Ring> interiorRings;
 
         const Ring* closestExterior = nullptr;
 
         bool isIn( const Ring& other ) const
         {
-            // FIXME
-            // This could probably be optimized by avoiding copies
-            // (and a dependency to OGR/GEOS)
-
-            assert( other.points.size() >= 4 );
-            Point p = points.front();
-
-            OGRLinearRing r;
-            for ( const auto& pt : other.points ) {
-                r.addPoint( pt.x, pt.y );
-            }
-            OGRPolygon poly;
-            poly.addRing( &r );
-
-            OGRPoint toTest( p.x, p.y );
-
-            return toTest.Within( &poly ) != 0;
-        }
-
-        void checkInclusionWith( const Ring& other )
-        {
-            if ( isIn( other ) ) {
-                if ( closestExterior ) {
-                    if ( other.isIn( *closestExterior ) ) {
-                        closestExterior = &other;
+            // Check if this is inside other using the winding number algorithm
+            auto checkPoint = this->points.front();
+            int windingNum = 0;
+            auto otherIter = other.points.begin();
+            // p1 and p2 define each segment of the ring other that will be tested
+            auto p1 = *otherIter;
+            do {
+                otherIter++;
+                auto p2 = *otherIter;
```

Comment:
I believe there's a risk to dereference otherIter when its already at other.points.end(). The test at the end of this scope (`` } while (...)``) should probably be moved here